### DISCUSSION TITLE
`enqueued_at`/`created_at` as integers

### DISCUSSION THREAD

**Discussion by fatkodima** [sidekiq/sidekiq#6562] [General]
This is kind of micro optimisation, but wanted to share - maybe you will find this interesting.

I did a simple sidekiq profiling using `RUBY_YJIT_ENABLE=1 LATENCY=0 THREADS=1 bin/sidekiqload` to see how much time it takes to enqueue a bunch of jobs.

```
Created 500000 jobs in 5.882485 sec
```

```
     TOTAL    (pct)     SAMPLES    (pct)     FRAME
      1479  (26.3%)        1479  (26.3%)     Float#to_s
      2035  (36.2%)         556   (9.9%)     JSON::Ext::Generator::State.generate
       418   (7.4%)         418   (7.4%)     IO#wait_readable
      5095  (90.8%)         302   (5.4%)     Sidekiq::Client#push_bulk
       289   (5.1%)         289   (5.1%)     (sweeping)
```

As can be seen, a lot of time is spent in `Float#to_s`. Turned out this is for `enqueued_at` and `created_at` standard sidekiq added fields.

```
▶ redis-cli
127.0.0.1:6379> info memory
# Memory
used_memory:96987232
used_memory_human:92.49M
```

So I made a simple change
```diff
diff --git a/lib/sidekiq/client.rb b/lib/sidekiq/client.rb
index 2a4fb6a1..1472840a 100644
--- a/lib/sidekiq/client.rb
+++ b/lib/sidekiq/client.rb
@@ -269,7 +269,7 @@ module Sidekiq
           [at, Sidekiq.dump_json(hash)]
         })
       else
-        now = Time.now.to_f
+        now = Time.now.to_i
         grouped_queues = payloads.group_by { |job| job["queue"] }
         conn.sadd("queues", grouped_queues.keys)
         grouped_queues.each do |queue, grouped_payloads|
diff --git a/lib/sidekiq/job_util.rb b/lib/sidekiq/job_util.rb
index cff0f386..02caa331 100644
--- a/lib/sidekiq/job_util.rb
+++ b/lib/sidekiq/job_util.rb
@@ -58,7 +58,7 @@ module Sidekiq
       item["class"] = item["class"].to_s
       item["queue"] = item["queue"].to_s
       item["retry_for"] = item["retry_for"].to_i if item["retry_for"]
-      item["created_at"] ||= Time.now.to_f
+      item["created_at"] ||= Time.now.to_i
       item
     end
```

Then
```
Created 500000 jobs in 3.791412 sec
```

```
▶ redis-cli
127.0.0.1:6379> info memory
# Memory
used_memory:88910080
used_memory_human:84.79M
```

Many realworld payloads are simple ids, so as can be seen having floats for 2 fields adds some overhead for json dumping time and for redis memory.

That was for 500k jobs. But if we consider, for example, batching over 10m users table and enqueueing a job for each of them,
previously it will take `118s` to enqueue, now `76s`, and now will use `180mb` less memory in redis.

I wonder if float precision (esp. 7 digits precision) is practically useful or only seconds is enough.

**Comment by mperham** [comment]
to_i likely isn't enough precision but millis likely is. I do know that parsing String to Float can be expensive for CPU, can you see any difference in time? I've considered switching to integer millis more for that reason. Also, Floats are just not as reliable in JSON because of JS's 53-bit numbers.

**Reply by fatkodima** [reply]
The bechmarking script does only dumping. I made a benchmark to test the difference:
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "json"
  gem "benchmark-ips"
end

require 'benchmark/ips'
require "json"

time = Time.now
float = time.to_f
integer = time.to_i

float_to_s = float.to_s
integer_to_s = integer.to_s

Benchmark.ips do |x|
  x.report("dumping float") { JSON.dump(float) }
  x.report("dumping integer") { JSON.dump(integer) }
  x.compare!
end

Benchmark.ips do |x|
  x.report("parsing float") { JSON.load(float_to_s) }
  x.report("parsing integer") { JSON.load(integer_to_s) }
  x.compare!
end
```

```
Comparison:
     dumping integer:  2348549.6 i/s
       dumping float:   500667.4 i/s - 4.69x  slower

Comparison:
     parsing integer:  1360228.9 i/s
       parsing float:   759904.3 i/s - 1.79x  slower
```

**Reply by mperham** [reply]
I do think it would be useful to migrate sidekiq away from using any floats in JSON as they've proven to be unreliable in various cases.

**Reply by fatkodima** [reply]
So basically storing milliseconds since epoch as an integer?
Let me know if you want to implement yourself.

**Reply by mperham** [reply]
Yeah, I think that's reasonable. I'm off for the week, no laptop. If you want a fun holiday hack project, you are welcome to implement.
